pkg/cloud/services/gc/service.go

```diff
@@ -0,0 +1,76 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gc
+
+import (
+	"context"
+
+	"github.com/aws/aws-sdk-go/aws/arn"
+	"github.com/aws/aws-sdk-go/service/ec2/ec2iface"
+	"github.com/aws/aws-sdk-go/service/elb/elbiface"
+	"github.com/aws/aws-sdk-go/service/elbv2/elbv2iface"
+	"github.com/aws/aws-sdk-go/service/resourcegroupstaggingapi/resourcegroupstaggingapiiface"
+
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/scope"
+)
+
+// ResourceCleanupFunc is a function type to cleaning up resources for a specific AWS service type.
+type ResourceCleanupFunc func(ctx context.Context, resources []*AWSResource) error
+
+// AWSResource represents a resource in AWS.
+type AWSResource struct {
+	ARN  *arn.ARN
+	Tags map[string]string
+}
+
+// Service is used to perform operations against a tenant/workload/child cluster.
+type Service struct {
+	scope                 cloud.ClusterScoper
+	elbClient             elbiface.ELBAPI
+	elbv2Client           elbv2iface.ELBV2API
+	resourceTaggingClient resourcegroupstaggingapiiface.ResourceGroupsTaggingAPIAPI
+	ec2Client             ec2iface.EC2API
+	cleanupFuncs          []ResourceCleanupFunc
+}
+
+// NewService creates a new Service.
+func NewService(clusterScope cloud.ClusterScoper, opts ...ServiceOption) *Service {
+	svc := &Service{
+		scope:                 clusterScope,
+		elbClient:             scope.NewELBClient(clusterScope, clusterScope, clusterScope, clusterScope.InfraCluster()),
```

Comment:
This looks a bit silly. :D Also, this entity implement WAAAY too many interfaces.

Reply:
Thats the convention used when creating services in CAPA.

Reply:
Yeah, no worries. I just found it a bit weird to pass in the same thing a couple of times since it implements a bunch of interfaces. Feels like it's an overloaded service. Or pass it in once, and then the client can sort out what it needs.

But this is not part of this code. :)

Reply:
We should look at this again in the future and refactor.

Reply:
Did you want to capture it as an issue?

Reply:
Yeah I'll create one! :)